src/ripple/rpc/impl/RPCHelpers.h

```diff
@@ -241,7 +241,7 @@ extern beast::SemanticVersion const lastVersion;
 constexpr unsigned int apiInvalidVersion = 0;
 constexpr unsigned int apiVersionIfUnspecified = 1;
 constexpr unsigned int apiMinimumSupportedVersion = 1;
-constexpr unsigned int apiMaximumSupportedVersion = 1;
+constexpr unsigned int apiMaximumSupportedVersion = 2;
 constexpr unsigned int apiBetaVersion = 2;
```

Comment:
As I wrote over [on #4571](https://github.com/XRPLF/rippled/pull/4571#discussion_r1228663775), once a version of `rippled` is released with `apiMaximumSupportedVersion` is set to 2, then we are saying that API v2 is **Done**. Once that happens we can not legitimately make any further changes to API version 2. To the best of my knowledge, API version 2 is still in beta, under active development, and still changing. 

Now, if you (and #4571 and #4568) are saying that version 2 is done, let's talk about version 2 being done.

If you want a node to properly handle API v2 requests, the correct way to do that, for now, is with the `[beta_rpc_api]` config option. And if you want to do it in unit tests, set it up in the config similar to how it's done in https://github.com/XRPLF/rippled/blob/develop/src/test/rpc/AccountInfo_test.cpp#L209-L212, or just by setting `cfg.BETA_RPC_API=true;` directly.

Reply:
Thanks for the feedback. I'll fix this.

Reply:
Thanks. Please see also #4573.

Reply:
Hi @ximinez , [this](https://github.com/XRPLF/rippled/blob/develop/src/test/rpc/AccountInfo_test.cpp#L209-L222) seems a bit problematic. In the test, `api_version 2` is specified in the payload which means that when let's say `api_version 3` is released (ofc, not soon but whenever it does) - the test will no longer run on `api_version 3` unless this particular behaviour changes and we rewrite the test for it. However ideally, would we not want the new test to run for all versions >= 2 unless the particular behaviour is changed for the next versions ofc ? I may be wrong, but I just wanted to clarify.

Reply:
Also now since we are using beta_version and not changing the max supported version, it does not matter for this PR. But I found a weird behaviour when I changed `apiMaximumSupportedVersion ` version to 2. It immediately failed a bunch of tests because the unit-tests started running on `apiMaximumSupportedVersion` by default which is inconsistent with [this](https://github.com/XRPLF/rippled/blob/beb0904a322bcaef777bedfe17c92c2d2f56e916/src/ripple/rpc/impl/RPCHelpers.h#L219-L239). I found [this](https://github.com/XRPLF/rippled/pull/4552/files#r1227259812). It is not important right now probably, but this might be a bug and we might need to fix this when `api_version 2` is done?

Reply:
This does not give me a good impression of our versioning scheme. Makes me think this is why version 2 has not released.

Do we have guidelines that answer these questions?

- How should we write code that is conditional on version and robust in the face of future versions?
- What do we call the set of versions that is tested? What do we call the set of versions that is enabled during operation?
- How do we configure which versions are in these sets?
- How can we ensure that all to-be-tested versions are actually tested? How can we ensure that code that is not conditional on version maintains its behavior across versions?

Reply:
It's the same as the way APIs are versioned in other API servers, unrelated to rippled.

You write code that is conditional by using conditions. When future versions are created, you have to update the code. Where updates are required depends on the changes being made in the API. The name for the set of versions that is tested is decided the same way as all other names in the code: propose something, and let other people propose alternatives. The same applies for the set of versions that are enabled during operation: you give it a name that makes sense to you, and is the best name you can think of, given limited time and knowledge. Similarly, you propose code that configures which versions are in these sets. And you have to write code that ensures all available versions are actually tested. The purpose of unit tests is to catch breakages when the API version changes, so it is good (and it is a sign of unit tests serving their intended purpose) when tests break in the face of breaking changes in the API. At that point, you update the tests so that they test the previous API version, and the current one.

If you have any other answers to your questions, go ahead and and share them.

Reply:
@ximinez gave some good answers in our internal Slack:

> 2. You mean like a name? I have no idea. As I said above, the tests should be over minimum to beta, and the set that's enabled during operation (by default) is minimum to maximum.
> 3. They are defined by the variables in `RPCHelpers.h`.
> 4. My thinking is that once [#4573](https://github.com/XRPLF/rippled/pull/4573) is merged, any tests that care about versions should loop from `apiMinimumSupportedVersion` to `apiBetaVersion` and run the tests across all versions. The tests would then have to be written with conditional code to handle the expected results across all versions. This is different than how a lot of the existing tests are written where there's one function for v1 and one for v2, which I agree with you is not sustainable.

He did not answer the first question in my opinion. I think it should be done with zero version _equalities_, only version _inequalities_. In fact, I think we can prescribe a pattern to be strictly-enforced:

```cpp
if (version < FEATURE_X_VERSION) {
  // old behavior
} else {
  // new behavior
}

if (version >= FEATURE_X_VERSION) {
  // new behavior where old behavior was just "do nothing"
}

if (version < FEATURE_X_VERSION) {
  // old behavior where new behavior is "stop doing that"
}
```